x-pack/solutions/security/plugins/security_solution/server/lib/siem_migrations/dashboards/task/agent/nodes/translate_panel/translate_panel.ts

```diff
@@ -0,0 +1,120 @@
+  return {
+    // Fan-in: the results of the individual panel translations are aggregated back into the overall dashboard state via state reducer.
+    node: async ({ index, ...nodeParams }) => {
+      try {
+        if (!nodeParams.parsed_panel.query) {
+          throw new Error('Panel query is missing');
+        }
+        const output = await translatePanelSubGraph.invoke(nodeParams);
+        return {
+          translated_panels: [
+            {
+              index,
+              data: output.elastic_panel ?? {},
+              translation_result: output.translation_result,
+            },
+          ],
+        };
+      } catch (err) {
+        return {
+          failed_panel_translations: [
+            {
+              index,
+              error_message: err.toString(),
+              details: err,
+            },
+          ],
+        };
+      }
+    },
+    // Fan-out: for each panel, Send translatePanel to be executed in parallel.
+    // This function needs to be called inside a `conditionalEdge`
+    conditionalEdge: (state: MigrateDashboardState) => {
+      const panels = state.parsed_original_dashboard.panels ?? [];
+      return panels.map((panel, i) => {
+        const resources = filterIdentifiedResources(
+          state.original_dashboard.vendor,
+          state.resources,
+          panel
+        );
+        const description = state.panel_descriptions[panel.id];
+        const translatePanelParams: TranslatePanelNodeParams = {
+          parsed_panel: panel,
+          description,
+          resources,
+          index: i,
+        };
+        return new Send('translatePanel', translatePanelParams);
+      });
+    },
+    subgraph: translatePanelSubGraph, // Only for the diagram generation
+  };
+};
+
+/**
+ * This function filters the stored resource data that have been received for the entire dashboard,
+ * and returns only the resources that have been identified for each specific panel query.
+ */
+function filterIdentifiedResources(
+  vendor: OriginalDashboardVendor,
+  resources: MigrationResources,
+  panel: ParsedPanel
+): MigrationResources {
+  const resourceIdentifier = new DashboardResourceIdentifier(vendor);
+  const identifiedResources = resourceIdentifier.fromQuery(panel.query);
+
+  const { macros, lookups } = identifiedResources.reduce<{ macros: string[]; lookups: string[] }>(
+    (acc, { type, name }) => {
+      if (type === 'macro') {
+        acc.macros.push(name);
+      } else if (type === 'lookup') {
+        acc.lookups.push(name);
+      }
+      return acc;
+    },
+    { macros: [], lookups: [] }
+  );
+  const filteredResources: MigrationResources = {};
+
+  const macro = resources.macro?.filter((m) => macros.includes(m.name));
+  if (macro?.length) {
+    filteredResources.macro = macro;
+  }
+  const lookup = resources.lookup?.filter((l) => lookups.includes(l.name));
+  if (lookup?.length) {
+    filteredResources.lookup = lookup;
```

Comment:
2 questions here:

1.  Why is this whole thing needed?  Can't we simply use `identifiedResources` Why do we need to check if it is already in our list of resources?
2. how do we use these resources?